inc/templates/namespace.php

```diff
@@ -364,263 +368,58 @@ function convert_blocks_to_components( array $blocks ): array {
 }
 
 /**
- * Hydrate components.
+ * Hydrate a JSON template into a REST response.
  *
- * @param array $components A list of components from a template.
- * @return array A hydrated array of components prepared for a REST response.
+ * @param array $data A JSON array of template data.
+ * @return array a data array ready for a REST response.
  */
-function hydrate_components( array $components ) {
+function hydrate_template( array $data ): array {
 
 	$hydrated = [];
 
-	foreach ( $components as $component ) {
-		// Create Object Instance to hydrate initial config values.
-		$component = setup_component( $component );
-
-		// Bail early if this isn't a WP_Irving\Component.
-		if ( ! $component instanceof Component ) {
-			break;
-		}
-
-		$template_data = hydrate_template_parts( $component );
+	foreach ( $data as $component ) {
+		// First, handle template part matching.
+		if (
+			isset( $component['name'] ) &&
+			0 === strpos( $component['name'], 'template-part/' )
+		) {
+			$template_data = hydrate_template_parts( $component );
 
-		// If the component is converted to template data,
-		// add hydrated components to the array and move on.
-		if ( ! empty( $template_data ) ) {
-			array_push( $hydrated, ...$template_data );
+			// If the component is converted to template data,
+			// add hydrated components to the array and move on.
+			if ( ! empty( $template_data ) ) {
+				array_push( $hydrated, ...$template_data );
+			}
 
-			// A little cleanup.
-			unset( $template_data );
-			break;
+			continue;
 		}
 
-		// Set up config from context.
-		$component->use_context( get_template_context() );
-
-		// Run hydration callback.
-		$component->do_callback();
-
-		// Update context values.
-		$component->provide_context( get_template_context() );
+		$component = new Component( $component['name'], $component );
```

Comment:
I'm unsure when this would even happen unless someone makes a mistake while writing a template file. I think I would rather throw an explicit warning/error message rather than hide/obscure the error so that it can more easily be identified and fixed.